subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -189,25 +192,35 @@ private String getFilename(Path dir) {
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
             String name = intern(file.getFileName().toString());
             if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
-                if (attrs == null) {
-                    throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
-                }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                FileSystemLocationSnapshot snapshot = getSnapshot(file, name, attrs);
+                builder.visit(snapshot);
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot getSnapshot(Path file, String name, BasicFileAttributes attrs) {
+            String absolutePath = intern(file.toString());
+            try {
+                HashCode hash = hasher.hash(file.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                FileMetadata fileMetadata = FileMetadata.from(attrs);
+                return new RegularFileSnapshot(absolutePath, name, hash, fileMetadata);
+            } catch (UncheckedIOException e) {
+                LOGGER.debug("Could not read file path '{}'.", file, e);
+                return new MissingFileSnapshot(absolutePath, name);
+            }
+        }
+
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
-            if (isNotFileSystemLoopException(exc) && isAllowed(file, file.getFileName().toString(), false, null, builder.getRelativePath())) {
-                throw new UncheckedIOException(String.format("Could not read path '%s'.", file), exc);
+            if (isNotFileSystemLoopException(exc)) {
+                String name = file.getFileName().toString();
+                if (isAllowed(file, name, Files.isDirectory(file), null, builder.getRelativePath())) {
+                    LOGGER.debug("Could not read file path '{}'.", file);
```

Comment:
this is for unreadable folders ... is `Files.isDirectory(file)` always true? because until now `isAllowed` was always called with `isDirectory = false`

Reply:
@lptr, @wolfs , was this a bug?

Reply:
What do you mean by "until now"? It is called with `isDirectory=true` here: https://github.com/gradle/gradle/blob/c53239b0722505b1f1516d6c822425799e3d0d0b/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L160

Reply:
That's a different line, I meant for `visitFileFailed`:
<img width="700" alt="Screen Shot 2019-07-01 at 12 42 22" src="https://user-images.githubusercontent.com/1841944/60426631-b52a7300-9bfd-11e9-8ca2-455400cd212e.png">

Reply:
Filtering for files (`isDirectory = false`) allows to actually ignore unreadable directories via `isAllowed`. If `isDirectory = true`, then `**` will cause the directory to be included so you won't be able to filter it...